### PR TITLE
feat: user access management

### PR DESCRIPTION
**Pull request by ypatil12**
# Overview

User Access Management (UAM) adds a core contract, the `PermissionController` as a dependency for the `DelegationManager` , `AllocationManager`, and `RewardsCoordinator` . We enable this functionality for AVSs and Operators but not for stakers. There is no support for the `AVSDirectory` as we plan on deprecating its functionality eventually. 

Operators & AVSs can have multiple admins for their account. This enables key rotation for operators. 

Operators cannot appoint an address to _register_ on its behalf.  In other words, the address they register with is their account and they can then transfer admin of that account in a subsequent tx. Similarly, the account of an AVS is the address they use to initialize state in the `AllocationManager`. 

# Added functionality

We add a `registerAsAVS` function as the entrypoint to become an AVS in the `AllocationManager`

# Edge Cases

- An admin can only be removed if there is at least 1 admin remaining after removal 

# TODOs

- [x] Tests with other contracts

# Delegation Enabled Functions

We do NOT add UAM support for `processClaim` (ie. claim rewards) as that function is used by stakers, operators, and (in RewardsV2) AVSs. There is already delegation for this function via the `setClaimerFor`. The downside here is that there can only be 1 claimer for an operator or AVS, as opposed to multiple for other functions. This seems acceptable for now, but open to hearing any concerns. 

## Operators

- `registerForOperatorSets`
- `deregisterFromOperatorSets`
- `modifyAllocations`
- `setAllocationDelay`
- `setClaimerFor`
- `modifyOperatorDetails`
- `undelegate`
- `updateOperatorMetadataURI`

## AVSs

- `deregisterFromOperatorSets`
- `slashOperator`
- `setAVSRegistrar`
- `updateAVSMetadataURI`
- `createOperatorSets`
- `addStrategiesToOperatorSet`
- `removeStrategiesFromOperatorSet`
- `createAVSRewardsSubmission`
- `setClaimerFor`

### CODE REVIEW

File: src/contracts/mixins/PermissionControllerMixin.sol [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        _;
+    }
+
+    /**
+     * @notice Checks if the caller is allowed to call a function on behalf of an account.
+     * @param account the account to check
+     * @dev `msg.sender` is the caller to check that can call the function on behalf of `account`.
+     * @dev Returns a bool, instead of reverting
+     */
+    function _checkCanCall(
```
Comment (stevennevins) [inline src/contracts/mixins/PermissionControllerMixin.sol:33]: pass in the delegate vs inferring

File: src/contracts/mixins/PermissionControllerMixin.sol [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        IPermissionController _permissionController
+    ) {
+        permissionController = _permissionController;
+    }
+
+    /// @notice Checks if the caller (msg.sender) can call on behalf of an account.
+    modifier checkCanCall(
+        address account
+    ) {
+        require(permissionController.canCall(account, msg.sender, address(this), msg.sig), InvalidPermissions());
```
Comment (stevennevins) [inline src/contracts/mixins/PermissionControllerMixin.sol:23]: use internal function and pass msg.sender as delegate

**Review by 0xClandestine — COMMENTED** [review]
Just some renaming suggestions.

File: src/contracts/permissions/PermissionController.sol [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+     */
+
+    /// @inheritdoc IPermissionController
+    function isAdmin(address account, address caller) public view returns (bool) {
+        if (_permissions[account].admins.length() == 0) {
+            // If the account does not have an admin, the caller must be the account
+            return account == caller;
+        } else {
+            // If the account has an admin, the caller must be an admin
+            return _permissions[account].admins.contains(caller);
```
Comment (8sunyuan) [inline src/contracts/permissions/PermissionController.sol:127]: this is done in constant lookup time so probably can just remove the length == 0 check. Remove the extra storage check for users who have admins set

File: src/contracts/permissions/PermissionController.sol [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    function setAdmin(address account, address admin) external onlyAdmin(account) {
+        // Add the admin to the account's admins
+        // If the admin is already set, the set will fail
+        EnumerableSet.AddressSet storage admins = _permissions[account].admins;
+        require(admins.add(admin), AdminAlreadySet());
+
+        emit AdminSet(account, admin);
+    }
+
+    function removeAdmin(address account, address admin) external onlyAdmin(account) {
```
Comment (8sunyuan) [inline src/contracts/permissions/PermissionController.sol:44]: So if you remove all admins, the original account becomes the admin again?
Is this desired? Or can we just prevent ever having 0 admins when you try to remove.

File: src/contracts/interfaces/IPermissionController.sol [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    /**
+     * @notice Set a delegate for a given account
+     * @param account to set delegate for
+     * @param delegate to set
+     * @param target to set delegate for
+     * @param selector to set delegate for
+     * @dev Only the admin of the account can set a delegate
+     */
+    function setDelegate(address account, address delegate, address target, bytes4 selector) external;
```
Comment (8sunyuan) [inline src/contracts/interfaces/IPermissionController.sol:56]: Is this getting renamed? As discussed before we should not confuse this with staker<>operator delegation.
I suggested something like `Entrustee` or `Entrant`. Entrant seems like it describes the actor who can enter specific function selectors.

File: src/contracts/permissions/PermissionControllerStorage.sol [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+
+import "../interfaces/IPermissionController.sol";
+
+abstract contract PermissionControllerStorage is IPermissionController {
+    using EnumerableSet for EnumerableSet.Bytes32Set;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    struct AccountPermissions {
```
Comment (8sunyuan) [inline src/contracts/permissions/PermissionControllerStorage.sol:12]: I quite like this storage design

File: src/contracts/interfaces/IPermissionController.sol (hunk 5 above)
Comment (stevennevins) [reply src/contracts/interfaces/IPermissionController.sol:56]: not a fan on `Entrant`

File: src/contracts/permissions/PermissionControllerStorage.sol (hunk 6 above)
Comment (stevennevins) [reply src/contracts/permissions/PermissionControllerStorage.sol:12]: lets go the rest of the way with this pattern and used salted storage location. Then we don't have to worry about storage gaps

File: src/contracts/permissions/PermissionController.sol [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    /**
+     *
+     *                         INTERNAL FUNCTIONS
+     *
+     */
+
+    /// @dev Encodes the target and selector into a single bytes32 value
+    function _encodeTargetSelector(address target, bytes4 selector) internal pure returns (bytes32) {
```
Comment (8sunyuan) [inline src/contracts/permissions/PermissionController.sol:99]: Would be good to have reversability unit test. Where you get the same starting target,selector values.

_encodeTargetSelector(target, selector) => bytes32 targetSelector
_decodeTargetSelector(targetSelector) => (target, selector)

File: src/contracts/permissions/PermissionController.sol [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+        // Remove the admin from the account's admins
+        // If the admin is not set, the remove will fail
+        require(admins.remove(admin), AdminNotSet());
+
+        emit AdminRemoved(account, admin);
+    }
+
+    /// @inheritdoc IPermissionController
+    function setDelegate(
+        address account,
```
Comment (stevennevins) [inline src/contracts/permissions/PermissionController.sol:56]: can we update to:

delegator

delegatee

File: src/contracts/permissions/PermissionController.sol [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+    function initialize() external initializer {}
+
+    /**
+     *
+     *                         EXTERNAL FUNCTIONS
+     *
+     */
+
+    /// @inheritdoc IPermissionController
+    function setAdmin(address account, address admin) external onlyAdmin(account) {
```
Comment (stevennevins) [inline src/contracts/permissions/PermissionController.sol:35]: addAdmin

isAdmin(account)

File: src/contracts/permissions/PermissionController.sol (hunk 8 above)
Comment (stevennevins) [reply src/contracts/permissions/PermissionController.sol:56]: addDelegate

File: src/contracts/permissions/PermissionController.sol [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+    /**
+     *
+     *                         INTERNAL FUNCTIONS
+     *
+     */
+
+    /// @dev Encodes the target and selector into a single bytes32 value
+    function _encodeTargetSelector(address target, bytes4 selector) internal pure returns (bytes32) {
+        return bytes32(abi.encodePacked(target, uint96(bytes12((selector)))));
```
Comment (stevennevins) [inline src/contracts/permissions/PermissionController.sol:100]: Pretty sure this could be simplified to
```
bytes32(abi.encodePacked(target, selector));
```
Decoding would be 

```
address target = address(uint160(uint256(targetSelector) >> 32));
bytes4 selector = bytes4(uint32(uint256(targetSelector));
```

or 
```
(address target, bytes4 selector) = abi.decode(bytes(targetSelector), (address, bytes4));
```

Comment (stevennevins) [reply src/contracts/permissions/PermissionController.sol:100]: without casting in the decode
```
(address target, bytes4 selector) = abi.decode(targetSelector, (address, bytes4));
```

File: src/contracts/permissions/PermissionController.sol [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+        AccountPermissions storage permissions = _permissions[account];
+
+        bytes32 targetSelector = _encodeTargetSelector(target, selector);
+        require(permissions.delegatePermissions[delegate].contains(targetSelector), DelegateNotSet());
+
+        // Remove the delegate from the account's permissions
+        permissions.delegatePermissions[delegate].remove(targetSelector);
+        permissions.permissionDelegates[targetSelector].remove(delegate);
+
+        emit DelegateRemoved(account, delegate, target, selector);
```
Comment (stevennevins) [inline src/contracts/permissions/PermissionController.sol:89]: We should capture who added the permission since it might not be the `account`

Comment (stevennevins) [inline src/contracts/permissions/PermissionController.sol:89]: We should capture who `msg.sender` removed the delegate

File: src/contracts/permissions/PermissionController.sol [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+        AccountPermissions storage permissions = _permissions[account];
+
+        bytes32 targetSelector = _encodeTargetSelector(target, selector);
+        require(!permissions.delegatePermissions[delegate].contains(targetSelector), DelegateAlreadySet());
+
+        // Add the delegate to the account's permissions
+        permissions.delegatePermissions[delegate].add(targetSelector);
+        permissions.permissionDelegates[targetSelector].add(delegate);
+
+        emit DelegateSet(account, delegate, target, selector);
```
Comment (stevennevins) [inline src/contracts/permissions/PermissionController.sol:70]: same here for who removed the delegate

File: src/contracts/permissions/PermissionController.sol [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+            return admin;
+        } else {
+            return _permissions[account].admins.values();
+        }
+    }
+
+    /// @inheritdoc IPermissionController
+    function canCall(address account, address caller, address target, bytes4 selector) external view returns (bool) {
+        return isAdmin(account, caller)
+            || _permissions[account].delegatePermissions[caller].contains(_encodeTargetSelector(target, selector));
```
Comment (stevennevins) [inline src/contracts/permissions/PermissionController.sol:147]: Can we make a view function

`isPermissioned(address account, address caller, address target, bytes4 selector);`

that does the logic after `||`

File: src/contracts/interfaces/IPermissionController.sol (hunk 5 above)
Comment (ypatil12) [reply src/contracts/interfaces/IPermissionController.sol:56]: How about agent??

File: src/contracts/mixins/PermissionControllerMixin.sol [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    /**
+     * @notice Checks if the caller is allowed to call a function on behalf of an account.
+     * @param account the account to check
+     * @dev `msg.sender` is the caller to check that can call the function on behalf of `account`.
+     * @dev Returns a bool, instead of reverting
+     */
+    function _checkCanCall(
+        address account
+    ) internal returns (bool) {
+        return permissionController.canCall(account, msg.sender, address(this), msg.sig);
```
Comment (wadealexc) [inline src/contracts/mixins/PermissionControllerMixin.sol:36]: Can this return `msg.sender == account || <external call>`? That way a short-circuit will prevent an external call for what feels like a likely common case.

File: src/contracts/permissions/PermissionController.sol (hunk 8 above)
Comment (ypatil12) [reply src/contracts/permissions/PermissionController.sol:56]: So we don't want to overload this term. How about delegate?

File: src/contracts/permissions/PermissionController.sol (hunk 4 above)
Comment (wadealexc) [inline src/contracts/permissions/PermissionController.sol:44]: Needs `inheritdoc`

Comment (wadealexc) [reply src/contracts/permissions/PermissionController.sol:44]: Or as Michael said prevent having 0 when you remove

File: src/contracts/permissions/PermissionController.sol [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+    /// @dev Decodes the target and selector from a single bytes32 value
+    function _decodeTargetSelector(
+        bytes32 targetSelector
+    ) internal view returns (address, bytes4) {
+        address target = address(uint160(uint256(targetSelector) >> 96));
+        // The selector is in the lower 32 bits of the targetSelector
+        bytes4 selector = bytes4(uint32(uint256(targetSelector) >> 64));
+
+        return (target, selector);
+    }
```
Comment (wadealexc) [inline src/contracts/permissions/PermissionController.sol:112]: If we are using shifts to decode, we should consider using shifts to encode. Makes it easier to reason that they're inverse operations.

File: src/contracts/permissions/PermissionController.sol (hunk 8 above)
Comment (ypatil12) [reply src/contracts/permissions/PermissionController.sol:56]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/870/commits/cf26f4eb29af4a89763b2e26b99802c8fd04d090

File: src/contracts/interfaces/IPermissionController.sol (hunk 5 above)
Comment (ypatil12) [reply src/contracts/interfaces/IPermissionController.sol]: Went with appointee: https://github.com/Layr-Labs/eigenlayer-contracts/pull/870/commits/cf26f4eb29af4a89763b2e26b99802c8fd04d090

File: src/contracts/mixins/PermissionControllerMixin.sol [hunk 16]
```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.0;
+
+import "../interfaces/IPermissionController.sol";
+
+abstract contract PermissionControllerMixin {
```
[the remark this answers is not in the document]
Comment (ypatil12) [reply src/contracts/mixins/PermissionControllerMixin.sol:6]: We do Mixin in the SignatureChecker too? You mean `PermissionControlledMixin`

File: src/contracts/mixins/PermissionControllerMixin.sol [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+    IPermissionController public immutable permissionController;
+
+    constructor(
+        IPermissionController _permissionController
+    ) {
+        permissionController = _permissionController;
+    }
+
+    /// @notice Checks if the caller (msg.sender) can call on behalf of an account.
+    modifier checkCanCall(
```
[the remark this answers is not in the document]
Comment (ypatil12) [reply src/contracts/mixins/PermissionControllerMixin.sol:20]: Hmm, to me `checkCanCall` is still the simplest to read

File: src/contracts/permissions/PermissionControllerStorage.sol (hunk 6 above)
Comment (ypatil12) [reply src/contracts/permissions/PermissionControllerStorage.sol:12]: Prefer not using salted location for consistency with storage layouts across core

File: src/contracts/permissions/PermissionController.sol (hunk 9 above)
Comment (ypatil12) [reply src/contracts/permissions/PermissionController.sol]: Changed function to addAdmin. Kept onlyAdmin modifier given we have a function with same name: https://github.com/Layr-Labs/eigenlayer-contracts/pull/870/commits/ccb8c89bf32175dc20f9d2b77f179db465ab9f98

File: src/contracts/permissions/PermissionController.sol (hunk 4 above)
Comment (ypatil12) [reply src/contracts/permissions/PermissionController.sol:45]: Let's prevent 0. We want to specifically enable permanent key rotation

File: src/contracts/permissions/PermissionControllerStorage.sol (hunk 6 above)
Comment (ypatil12) [reply src/contracts/permissions/PermissionControllerStorage.sol:12]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/870/commits/eac8bbe1116ca3cef86eaada1f9d6d666aff49f7

File: src/contracts/permissions/PermissionController.sol (hunk 4 above)
Comment (ypatil12) [reply src/contracts/permissions/PermissionController.sol:45]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/870/commits/eac8bbe1116ca3cef86eaada1f9d6d666aff49f7

File: src/contracts/permissions/PermissionController.sol (hunk 15 above)
Comment (ypatil12) [reply src/contracts/permissions/PermissionController.sol]: Reads cleaner: https://github.com/Layr-Labs/eigenlayer-contracts/pull/870/commits/cd629bed6e808752567b701a1fc607d0f9281d14

File: src/contracts/permissions/PermissionController.sol (hunk 4 above)
Comment (ypatil12) [reply src/contracts/permissions/PermissionController.sol:45]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/870/commits/ccb8c89bf32175dc20f9d2b77f179db465ab9f98

File: src/contracts/permissions/PermissionController.sol (hunk 11 above)
Comment (ypatil12) [reply src/contracts/permissions/PermissionController.sol]: Hm, not seeing the use case here

File: src/contracts/mixins/PermissionControllerMixin.sol (hunk 14 above)
Comment (ypatil12) [reply src/contracts/mixins/PermissionControllerMixin.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/870/commits/efe8e2e4ad33a3a1341ba9d62fe65d5e0afcff05

File: src/contracts/permissions/PermissionController.sol (hunk 13 above)
Comment (ypatil12) [reply src/contracts/permissions/PermissionController.sol]: Not sure why we would want to call a subset of this function?

File: src/contracts/permissions/PermissionController.sol (hunk 10 above)
Comment (ypatil12) [reply src/contracts/permissions/PermissionController.sol]: Don't think any of those work??

`bytes32(abi.encodePacked(target, selector))`

The encoded would be: [160 bits target][32 bits selector][64 bits padding]

`bytes4 selector = bytes4(uint32(uint256(targetSelector));`

Ends up being all zeros because we're getting the least significant 32 bits of the encoded value. 

Can't do abi.decode since it depends on bytes and would rather not do the bytes.concat(bytes32).

File: src/contracts/mixins/PermissionControllerMixin.sol (hunk 1 above)
Comment (ypatil12) [reply src/contracts/mixins/PermissionControllerMixin.sol:33]: Will do for the function, but modifier is much cleaner without

Comment (ypatil12) [reply src/contracts/mixins/PermissionControllerMixin.sol:33]: Can do for the function, but the modifier is much cleaner without. Still think we should pass in the appointee (formerly called delegate)?

File: src/contracts/permissions/PermissionController.sol (hunk 7 above)
Comment (ypatil12) [reply src/contracts/permissions/PermissionController.sol:103]: Added basic reversibility: https://github.com/Layr-Labs/eigenlayer-contracts/pull/870/commits/343b44e11fe3b3824941fbb6fe05a19307b93c25

**Comment by ypatil12** [comment]
> If an admin is set, neither the isAdmin or canCall methods give permission when account == caller.

Since granting permissions is not 2-step, it's possible to brick your delegation just by setting your first admin to an account that can't do anything.

Are we aware of this limitation? Could we consider allowing account == caller to act as an admin/universal permissions?

Yes we are aware. The reason why we don't allow `account == caller` to act as a universal admin is that we want to enable complete key rotation and doing so results in the original account key having perms. 

Adding a 2-step makes sense

File: src/contracts/core/AllocationManager.sol [hunk 18]
```diff
@@ -536,6 +536,11 @@ contract AllocationManager is
      *
      */
 
+    /// @inheritdoc IAllocationManager
+    function getAVSSetCount(address avs) external view returns (uint256) {
+        return _operatorSets[avs].length();
+    }
```
Comment (wadealexc) [inline src/contracts/core/AllocationManager.sol]: `getOperatorSetCount`

**Review by wadealexc — APPROVED** [review]
Looks good except for the name. Approving so you can merge after changing!

File: src/contracts/core/AllocationManager.sol (hunk 18 above)
Comment (ypatil12) [reply src/contracts/core/AllocationManager.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/870/commits/c24a46b0b9e17eda0b254b59680b9babf143a751

File: src/contracts/core/DelegationManager.sol
[the remark this answers is not in the document]
Comment (ypatil12) [reply src/contracts/core/DelegationManager.sol:1]: Yup, very good call out. I'll add this to the ELIP

[the remark this answers is not in the document]
Comment (ypatil12) [reply src/contracts/core/DelegationManager.sol:1]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/870/commits/a1c843d33f21478f4c16ffca7388d5a3fd66ce13
